sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -813,12 +839,34 @@ public String toString() {
     }
   }
 
-  private static class BigtableWriterFn
+  @VisibleForTesting
+  static class BigtableWriterFn
       extends DoFn<KV<ByteString, Iterable<Mutation>>, BigtableWriteResult> {
 
-    BigtableWriterFn(BigtableConfig bigtableConfig) {
+    protected final Counter cumulativeThrottlingMilliseconds =
+        Metrics.counter(BigtableWriterFn.class, "throttling-msecs");
+    private static Object metricLock = new Object();
+    private static long lastAggregatedThrottleTime = 0;
+    private ResourceStatsSupplier statsSupplier;
+
+    BigtableWriterFn(BigtableConfig bigtableConfig, ResourceStatsSupplier statsSupplier) {
       this.config = bigtableConfig;
       this.failures = new ConcurrentLinkedQueue<>();
+      this.statsSupplier = statsSupplier;
+    }
+
+    public static BigtableWriterFn create(BigtableConfig bigtableConfig) {
+      if (bigtableConfig.getProjectId() == null || bigtableConfig.getInstanceId() == null) {
+        return new BigtableWriterFn(
+            bigtableConfig,
+            new ResourceStatsSupplierImpl(
+                bigtableConfig.getBigtableOptions().getProjectId(),
+                bigtableConfig.getBigtableOptions().getInstanceId()));
```

Comment:
I think this breaks templates.

There are 2 phases in a dataflow job: graph construction and execution. During construction, all of the DoFns are instantiated, but none of the ValueProviders are populated. Then during execution all of the ValueProviders are populated and DoFns are executed.

Here you are trying to resolve ValueProviders during construction time, which will break the template mechanism.
you can do one of 2 things:

1. pass value provider for the instance name (the valueprovider will take BigtableConfig as the input)
2. Pass the BigtableConfig and resolve the instance name later